tests/reg_tests/test_actuator.py

```diff
@@ -0,0 +1,814 @@
+
+        # this is the force computed by the momentum integration directly from CFD
+        # just sum these up, the forces contain the correct normals from CFD
+        cfd_force = fp_o + fm_o + fp_i + fm_i
+
+        # The low accuracy is because the intgrated quantities don't have a lot of precision
+        np.testing.assert_allclose(my_force, az_force, rtol=1e-3)
+        np.testing.assert_allclose(cfd_force, az_force, rtol=1e-3)
+
+        ##################
+        # TEST POWER ADDED
+        ##################
+
+        # this is the integration done in the AZ
+        az_power = funcs[self.ap.name + "_flowpower_az"]
+
+        # this is the energy balance of the control volume.
+        # Cp of air is taken as 1004.5 J/kg
+        my_power = 1004.5 * (mdot_o * ttot_o - mdot_i * ttot_i)
+
+        # the tolerance is slightly worse but not terrible
+        np.testing.assert_allclose(my_power, az_power, rtol=1.5e-3)
+
+    def test_actuator_heat_flux(self):
+        "Tests if the correct amount of heat is added to the flow by the actuator"
+
+        # set the az heat
+        az_heat = 1e5
+        # need to set all dvs because training may re-use leftover dvs from a previous test
+        self.ap.setDesignVars({"thrust": 0.0, "heat": az_heat})
+
+        # we have to rerun because restart file was ran with a different thrust value
+        self.CFDSolver(self.ap)
+        funcs = {}
+        self.CFDSolver.evalFunctions(self.ap, funcs)
+
+        # negate mdot out because of the normal, mdot in is already positive
+        mdot_i = funcs[self.ap.name + "_mdot_in"]
+        mdot_o = -funcs[self.ap.name + "_mdot_out"]
+
+        vx_i = funcs[self.ap.name + "_mavgvx_in"]
+        vx_o = funcs[self.ap.name + "_mavgvx_out"]
+
+        area_i = funcs[self.ap.name + "_area_in"]
+        area_o = funcs[self.ap.name + "_area_out"]
+
+        aavgps_i = funcs[self.ap.name + "_aavgps_in"]
+        aavgps_o = funcs[self.ap.name + "_aavgps_out"]
+
+        ttot_i = funcs[self.ap.name + "_mavgttot_in"]
+        ttot_o = funcs[self.ap.name + "_mavgttot_out"]
+
+        # also get the pressure and momentum forces directly from CFD
+        fp_i = funcs[self.ap.name + "_forcexpressure_in"]
+        fp_o = funcs[self.ap.name + "_forcexpressure_out"]
+
+        fm_i = funcs[self.ap.name + "_forcexmomentum_in"]
+        fm_o = funcs[self.ap.name + "_forcexmomentum_out"]
+
+        #####################
+        # TEST MOMENTUM ADDED
+        #####################
+
+        # this is the analytical force based on primitive values (like mdot, ps etc)
+        my_force = mdot_o * vx_o + aavgps_o * area_o - (mdot_i * vx_i + aavgps_i * area_i)
+
+        # this is the force computed by the momentum integration directly from CFD
+        # just sum these up, the forces contain the correct normals from CFD
+        cfd_force = fp_o + fm_o + fp_i + fm_i
+
+        # the true value is 0, so we add 100 on both sides to have a reasonable comparison
```

Comment:
This seems like a hack to me. 
Why not just use an absolute tolerance?

Reply:
I don't know how to fix this, and I am probably being dumb but I also could not find a direct answer online. The issue is because the target is exactly zero. Then the relative error ends up being infinite:
```
AssertionError: 
Not equal to tolerance rtol=0.001, atol=0.001

Mismatched elements: 1 / 1 (100%)
Max absolute difference: 0.07964251
Max relative difference: inf
 x: array(0.079643)
 y: array(0.)
```

Do you have any suggestions on how to fix this? I tried setting a rtol of `np.inf` but still fails. I am not sure if there is a version that just checks atol. Swapping the order of value and target also does not help, and that's actually not a great solution in general imo.

What I can do is set the target to the small 0.07964251 value. This is due to the accumulation of integration errors and flow convergence in ADflow. If I compare against this (nonzero) target, then both atol and rtol will work. Or I could add a finite number to both sides (100, or 1 or whatever). Let me know what you think is the best solution here. I am leaving this test in a broken state so we dont merge until this is fixed.

Reply:
I think the `inf` is expected, it's just that the `atol` is too tight.

Reply:
The function `numpy.testing.assert_allclose` that you used should do exactly what you want, is that not the case? The [docstrings](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_allclose.html) say it should compare using `atol + rtol * abs(desired)` so if `desired=0` then it should use just `atol`. Maybe you need a looser `atol` since the actual tolerance is around 0.08.

However, if you expect the integration error to be around 0.08 then that may be a better value to use. That will ensure a tighter tolerance and check if in the future this value changes by a small amount.

Reply:
As you correctly pointed out, I am dumb. I set an absolute tolerance of 0.08, and I still compare against a reference of 0. Doing it this was seemed more straightforward for me rather than explaining why the target value is a specific number. I think those type of checks should be done with a comparison to the reference.json file so that they can be trained. 

I think checking with this relatively loose tolerance is okay, because I check all functionals in the test "test_actuator_thrust_and_heat" against the reference values in https://github.com/mdolab/adflow/blob/b3469288f03f99b03032621f01990d836f137513/tests/reg_tests/refs/actuator_tests.json with `rtol=1e-12` and `atol=1e-12`. In summary, I think the tests are good to go imo if you are happy with the final modification.

Reply:
I wanted to keep the comparison to zero just to demonstrate we expect a value of zero here, rather than some other number. Also, I test pretty much everything that touches the actuator zone so I am pretty comfortable with this individual one.

Reply:
I am also fine with this because it basically says that the integration error should not increase past 0.08